### PR TITLE
[RFC] Fix fold bug when switching from a terminal buffer

### PR DESCRIPTION
**Pull request by rdnetto**
When switching from a terminal buffer to a file using `foldmethod=expr`, Neovim jumps to line 19 (or sometimes 20).

This appears to have been introduced by commit e1079c2a for #5351, which I believe unintentionally deleted an if-statement that I have re-added.

(CC: @Shougo)

### CODE REVIEW

**Comment by justinmk** [comment]
Thanks. Can you add a test?

**Comment by rdnetto** [comment]
I've tried to create a test for it by duplicating the `foldmethod=expression` test, but I can't reproduce the problem with the folding function it uses. The one I'm able to reproduce it with is https://github.com/tkonolige/vim-haskellFold.git, but I'm not able to identify which part of it is triggering the behaviour. Would you mind taking a look at it and seeing if there's anything that stands out to you? The problem can be reproduced with [this file](https://github.com/rdnetto/powerline-hs/blob/master/src/Main.hs).

**Comment by bohrshaw** [comment]
@rdnetto There may be a test case in #6769.

**Comment by justinmk** [comment]
This doesn't seem like the right solution because we plan to support folds in terminal buffers.  `refresh_timer_cb` must have been triggered,  but why is `wp->w_buffer->terminal` non-NULL here?

**Comment by rdnetto** [comment]
Thanks @bohrshaw, looks like that test case works. I'll update the PR to use it.

@justinmk without any familiarity with the code, my guess would be things are happening in the wrong order. At any rate, I think it would be better to just fix the current behaviour (once I've added a working test) which is impacting users, than to hold out for a perfect solution to enable functionality we don't yet have.

**Comment by rdnetto** [comment]
OK, this is weird. I can definitely reproduce that problem running the command in #6769 and verify that the fix works. However, when I try to write a test case for it, the test output doesn't match the behaviour I see when manually executing the instructions, and I'm at a loss as to why.

File: test/functional/legacy/045_folding_spec.lua [hunk 1]
```diff
@@ -190,6 +190,17 @@ describe('folding', function()
       0]])
   end)
 
+  it("Regression #6549", function()
+    feed_command('let g:markdown_folding = 1')
+    feed_command('term top')
+    feed_command('e 045_test_data.md')
```
Comment (bohrshaw) [inline test/functional/legacy/045_folding_spec.lua]: `e + 045_test_data.md` may work.

File: test/functional/legacy/045_folding_spec.lua [hunk 2]
```diff
@@ -190,6 +190,17 @@ describe('folding', function()
...
+
+    feed_command('sleep 2')
```
Comment (bohrshaw) [inline test/functional/legacy/045_folding_spec.lua]: In my testing, the current mode at this moment is Insert Mode. And the cursor would only jump when in Normal Mode.

**Comment by teto** [comment]
rebase ?

**Pull request closed without merging**